File: harness/semgrep/fixtures/failopen/store.good.ts
// GOOD: a shared store is required; mis-configuration fails closed. The
// in-memory impl is only ever constructed behind an explicit opt-in flag, and
// the binding is not named like a replay/session store default.

import { createMemorySessionStore } from './store';

interface EngineConfig {
    store?: SessionStore;
    allowSingleProcessMemoryStore?: boolean;
}

export function createEngine(config: EngineConfig) {
    if (!config.store && !config.allowSingleProcessMemoryStore) {
        throw new Error('a shared SessionStore is required outside single-process deploys');
    }
    const backing = config.store ?? explicitSingleProcess(config);
    return { store: backing };
}

// Explicit opt-in helper — the memory store is returned from a clearly named
// function, not silently defaulted in a `?? createMemory...()` fallback.
function explicitSingleProcess(config: EngineConfig): SessionStore {
    return createMemorySessionStore();
}

File: harness/semgrep/fixtures/failopen/store.good.go
package server

// GOOD: a shared/persistent store is REQUIRED. No silent in-memory fallback —
// mis-configuration fails CLOSED (error) rather than fanning out per-process.

import "errors"

type Config struct {
	Store SessionStore
}

func NewServer(config Config) (*Server, error) {
	if config.Store == nil {
		return nil, errors.New("a shared SessionStore is required outside single-process deploys")
	}
	return &Server{store: config.Store}, nil
}

// GOOD: an in-memory store is fine when it is an explicit opt-in, not a
// silent nil-fallback. The variable here is not gated on a nil config field.
func newSingleProcessMethod() *Method {
	store := NewMemoryChannelStore()
	return &Method{store: store}
}
